### PR TITLE
feat: kwok cloud provider

### LINKED ISSUE #895: Karpenter KwoK Cloud Provider
### Description

**What problem are you trying to solve?**
Developers that want to make contributions to this repo must build and install with a cloud provider specific installation. This can be costly for users to develop and test their features, and may require maintainers to do that testing for them. Additionally, since this comes with Cloud Provider specific details, users contributing just to kubernetes-sigs/karpenter have to be aware of how the specific CP implementation may have decided to reason about general concepts. 

Creating a cloud provider using [KwoK nodes](https://github.com/kubernetes-sigs/kwok/tree/main/pkg) won't incur costs for users that need to test and validate their changes.

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### PR DESCRIPTION
**Pull request by njtran**
Fixes #895

**Description**
This adds the KwoK Cloud Provider, and moves some of the existing fake cloud provider. This also adds in the Karpenter cloud provider agnostic charts in, only to be used for installing the KwoK cloud provider.

**How was this change tested?**
following the instructions listed in the PR. 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: pkg/cloudprovider/fake/controller-kwok/charts/templates/aggregate-clusterrole.yaml
Comment (jonathan-innis) [inline pkg/cloudprovider/fake/controller-kwok/charts/templates/aggregate-clusterrole.yaml:1]: How much of this chart can be dropped since there is no need for backwards compatability? e.g. Can we drop the webhooks, the configmap logging, etc.?

File: pkg/cloudprovider/fake/controller-kwok/kwok/README.md
Comment (jonathan-innis) [inline pkg/cloudprovider/fake/controller-kwok/kwok/README.md:1]: This directory structure feels strange to me. I wonder if we should just think about doing kwok at the top-level and putting all of the stuff related to the kwok cloudprovider in that bucket. Makes more sense there as well since we are almost considering it a different project. It probably deserves its own go.mod too since I think we would optimistically like to move this into `kubernetes-sigs/karpenter-provider-kwok` down the line and just pull it out of here altogether.

File: pkg/cloudprovider/fake/controller-kwok/main.go
Comment (jonathan-innis) [inline pkg/cloudprovider/fake/controller-kwok/main.go:43]: nit: cut this line

File: pkg/cloudprovider/fake/controller-kwok/main.go
Comment (jonathan-innis) [inline pkg/cloudprovider/fake/controller-kwok/main.go:44]: Is this still needed?

File: pkg/cloudprovider/fake/controller-kwok/main.go
Comment (jonathan-innis) [inline pkg/cloudprovider/fake/controller-kwok/main.go:1]: I'm assuming that there is some follow-up work that is getting proposed here that would eventually see us pushing the chart and an image of the cloudprovider to some registry that everyone can access without having to checkout this repo?

File: Makefile
Comment (jonathan-innis) [inline Makefile:2]: Should we pushing to an AWS repo since this image should exist out in the open? We should probably talk to SIG Autoscaling or other release SIGs around where we should be storing the artifacts that are published from this repo

Comment (jonathan-innis) [reply Makefile:2]: There's a shorter-term question that I have here which is: Can we just releae the code for the provider and allow users to do a `make run` which runs the binary against their repo but doesn't install it directly as a deployment on their cluster and then do the release piece as a fast-follow once we decide where the artifacts land

File: go.mod
Comment (jonathan-innis) [inline go.mod:9]: Why did this dep get added?

File: hack/install-kwok.sh
Comment (jonathan-innis) [inline hack/install-kwok.sh:1]: Is this script based off of anything? Or did we write it ourselves? If it's based on some existing script, we should consider linking that script here.

File: pkg/cloudprovider/fake/controller-kwok/charts/README.md
Comment (jonathan-innis) [inline pkg/cloudprovider/fake/controller-kwok/charts/README.md:13]: Should we reference the AWS docs for this right now?

File: pkg/cloudprovider/fake/controller-kwok/charts/README.md.gotmpl
Comment (jonathan-innis) [inline pkg/cloudprovider/fake/controller-kwok/charts/README.md.gotmpl:19]: Some of these settings should be dropped since they are specific to AWS

File: pkg/cloudprovider/fake/controller-kwok/charts/crds/karpenter.sh_nodeclaims.yaml
Comment (jonathan-innis) [inline pkg/cloudprovider/fake/controller-kwok/charts/crds/karpenter.sh_nodeclaims.yaml:1]: This path doesn't look correct since we are now a few directories deeper with this chart

File: pkg/cloudprovider/fake/controller-kwok/charts/crds/karpenter.sh_nodepools.yaml
Comment (jonathan-innis) [inline pkg/cloudprovider/fake/controller-kwok/charts/crds/karpenter.sh_nodepools.yaml:1]: Same comment here about the symlink

File: pkg/cloudprovider/fake/controller-kwok/kwok/README.md
Comment (jonathan-innis) [reply pkg/cloudprovider/fake/controller-kwok/kwok/README.md:1]: Thinking about this a little more, I actually think that moving it into a separate repo long-term may not be the right move. We probably want to couple updates to the logic in our libraries with the update of the kwok cloudprovider for testing and coordinating those changes across two repos is too high of a burden compared to the benefit of decoupling. I still think we might want to shift around the directory structure here since it might be confusing to see a full chart hidden in the cloudprovider directory here.

File: pkg/cloudprovider/fake/controller-kwok/charts/templates/aggregate-clusterrole.yaml
Comment (njtran) [reply pkg/cloudprovider/fake/controller-kwok/charts/templates/aggregate-clusterrole.yaml:1]: Good callout, I hadn't thought about what was actually needed vs what was core and what wasn't. We can probably drop webhooks and configmap logging, but can take another pass at it in a separate PR, if that makes it easier to review.

File: pkg/cloudprovider/fake/controller-kwok/kwok/README.md
Comment (njtran) [reply pkg/cloudprovider/fake/controller-kwok/kwok/README.md:1]: Agree with not a separate repo. I think we had discussed this in the past, where this is tightly coupled with Karpenter, and may not make sense as a standalone project. 

How do you want to see this differently? in a top level cmd folder like in aws/karpenter-provider-aws?

File: pkg/cloudprovider/fake/controller-kwok/main.go
Comment (njtran) [reply pkg/cloudprovider/fake/controller-kwok/main.go:1]: I see value in this as a developer tool, and I see additional value in users who just want to try out Karpenter. I'm not sure if we this second use-case is as common, as new Karpenter users should probably test on a real cloud provider. Thoughts on why you think we should be pushing this up to some registry?

File: go.mod
Comment (njtran) [reply go.mod:9]: docker/namesgenerator

File: Makefile
Comment (njtran) [reply Makefile:2]: This was purely for development purposes as an example, not an actual repo where we're storing/publishing artifacts. Users should be able to run `make run` or `make apply` to fit however they want to test upstream Karpenter. Not convinced about the value of doing a release on this though.

File: hack/install-kwok.sh
Comment (njtran) [reply hack/install-kwok.sh:1]: I wrote it myself

File: pkg/cloudprovider/fake/controller-kwok/charts/README.md
Comment (njtran) [reply pkg/cloudprovider/fake/controller-kwok/charts/README.md:13]: I just copied this from the charts in the AWS repo. I think the AWS getting started guide comes with a lot more than necessary for this, what are you thinking is not enough/too much to maintain with this?

File: pkg/cloudprovider/fake/controller-kwok/charts/README.md.gotmpl
Comment (njtran) [reply pkg/cloudprovider/fake/controller-kwok/charts/README.md.gotmpl:19]: Good call, I hadn't fixed the charts yet.

File: pkg/cloudprovider/fake/controller-kwok/charts/crds/karpenter.sh_nodeclaims.yaml
Comment (njtran) [reply pkg/cloudprovider/fake/controller-kwok/charts/crds/karpenter.sh_nodeclaims.yaml:1]: Good catch.

File: pkg/cloudprovider/fake/controller-kwok/main.go
Comment (jonathan-innis) [reply pkg/cloudprovider/fake/controller-kwok/main.go:1]: > Thoughts on why you think we should be pushing this up to some registry

There's potentially some cool use-cases that we could support here to run kwok with karpenter when building other controllers where you need to mock a real environment with pod and node scaling. This is definitely a musing more than anything else.

File: Makefile
Comment (jonathan-innis) [reply Makefile:2]: Got it. Makes sense.

File: pkg/cloudprovider/fake/controller-kwok/kwok/README.md
Comment (njtran) [reply pkg/cloudprovider/fake/controller-kwok/kwok/README.md:1]: Think that's a good idea. Let me try this.

File: kwok/cloudprovider/cloudprovider.go
Comment (jonathan-innis) [inline kwok/cloudprovider/cloudprovider.go:116]: nit: we just say `aws` in the name for the AWS provider. I think the provider bit is implied. We should check to see what Azure is doing as well

File: kwok/cloudprovider/cloudprovider.go
Comment (jonathan-innis) [inline kwok/cloudprovider/cloudprovider.go:197]:
```suggestion
	// add the derived nodeclaim requirement labels
```

File: kwok/cloudprovider/cloudprovider.go
Comment (jonathan-innis) [inline kwok/cloudprovider/cloudprovider.go:192]:
```suggestion
	// start with labels on the nodeclaim
```

File: kwok/cloudprovider/cloudprovider.go
Comment (jonathan-innis) [inline kwok/cloudprovider/cloudprovider.go:215]: Should this be configurable? Seems odd that we hardcode 10 here?

File: kwok/cloudprovider/cloudprovider.go
Comment (jonathan-innis) [inline kwok/cloudprovider/cloudprovider.go:223]: nit: should this be a constant value?

File: kwok/cloudprovider/cloudprovider.go
Comment (jonathan-innis) [inline kwok/cloudprovider/cloudprovider.go:212]: nit: reference a constant value? Either in the package itself or in our own constants file?

File: kwok/cloudprovider/register.go
Comment (jonathan-innis) [inline kwok/cloudprovider/register.go:20]: Kwok is currently using `kwok.x-k8s.io`. Maybe worth a question to their team if we can extend off of it to use `karpenter.kwok.x-k8s.io`. This doesn't require formal approval from API since it's `x-k8s.io` but it seems odd that we would use `.sh` here when other `kwok` resources (and the website that they host themselves) doesn't use the `.sh` domain suffix

File: kwok/cloudprovider/register.go
Comment (jonathan-innis) [inline kwok/cloudprovider/register.go:21]: Do we need this?

File: kwok/cloudprovider/register.go
Comment (jonathan-innis) [inline kwok/cloudprovider/register.go:34]: Should all of this be commented out?

**Review by jonathan-innis — COMMENTED** [review]
Nice work! A bunch of comments but I'm really excited to start using this!

File: kwok/charts/templates/role.yaml
Comment (njtran) [reply kwok/charts/templates/role.yaml:42]: Yeah I think we should be able to remove. Good catch!

File: kwok/cloudprovider/README.md
Comment (njtran) [reply kwok/cloudprovider/README.md:3]: There's no `make delete` in this repo, but probably good to have. I'll add.

File: kwok/cloudprovider/README.md
Comment (njtran) [reply kwok/cloudprovider/README.md:49]: Nope, just added the kwok bit

File: Makefile
Comment (njtran) [reply Makefile:27]: Personally I like keeping it as a separate target in case a user wants to do it individually. WDYT ?

File: kwok/cloudprovider/cloudprovider.go
Comment (njtran) [reply kwok/cloudprovider/cloudprovider.go:47]: Yeah good call, I'l change it to the controller-runtime for parity .

File: kwok/cloudprovider/cloudprovider.go
Comment (njtran) [reply kwok/cloudprovider/cloudprovider.go:71]: Sounds good. Happy to go with the flow here

File: kwok/cloudprovider/cloudprovider.go
Comment (njtran) [reply kwok/cloudprovider/cloudprovider.go:215]: I think we can make it configurable in the future. i think of it as a detail people really shouldn't need to mess with or have to be aware of. 10 is enough to make it scalable, but can definitely adjust it in the future.

File: kwok/cloudprovider/register.go
Comment (njtran) [reply kwok/cloudprovider/register.go:20]: I can send a question to the maintainers over there to see what their thought is, but think we can change that really at any point.

Comment (jonathan-innis) [reply kwok/cloudprovider/register.go]: Sure, "all of these labels are alpha", "this is really only meant for dev and we can break you at any time"

File: kwok/cloudprovider/README.md
Comment (jonathan-innis) [reply kwok/cloudprovider/README.md]: From what I can tell, this is still missing the detail of how to stand-up an ECR repo to be able to just run `make apply` out of the box

File: kwok/cloudprovider/cloudprovider.go
Comment (jonathan-innis) [inline kwok/cloudprovider/cloudprovider.go:47]:
```suggestion
func NewCloudProvider(ctx context.Context, kubeClient client.Client, instanceTypes []*cloudprovider.InstanceType) *CloudProvider {
```

nit: a little weird that that the argument doesn't match the member field

File: kwok/cloudprovider/cloudprovider.go
Comment (jonathan-innis) [inline kwok/cloudprovider/cloudprovider.go:86]: Should wrap `NodeClaimNotFound` here too

File: kwok/main.go
Comment (jonathan-innis) [inline kwok/main.go:28]: Does this restricted set of label domains or well-known labels change what our CRDs should look like here? We have some custom logic that was written to inject certain patterns and CEL expressions into the NodePool/NodeClaim based on the Cloud Provider requirements. Take a look here for more detail: https://github.com/aws/karpenter-provider-aws/tree/main/hack/validation

File: kwok/cloudprovider/cloudprovider.go
Comment (jonathan-innis) [inline kwok/cloudprovider/cloudprovider.go:41]: nit: We already have a constants file, should these just go in there?

**Review by jonathan-innis — APPROVED** [review]
/lgtm
/approve

Let's get it started, ha, let's get it started in here 🎵

**Pull request merged**
